<h2 style="text-align: center; color: #1f4ed8;">SVM (Support Vector Machine)</h2>

### Problem Statement:  Classify raisins into one of the two categories,
1. Kecimen
1. Besni

### Dataset Citation
This dataset is used under citation guidelines from the original authors. For detailed study and dataset description, see the following references:

- **Citation**: Cinar, I., Koklu, M., & Tasdemir, S. (2020). Classification of Raisin Grains Using Machine Vision and Artificial Intelligence Methods. *Gazi Journal of Engineering Sciences, 6*(3), 200-209. DOI: [10.30855/gmbd.2020.03.03](https://doi.org/10.30855/gmbd.2020.03.03)
- **Dataset available at**: [Murat Koklu's Dataset Page](https://www.muratkoklu.com/datasets/)
- **Article download**: [DergiPark](https://dergipark.org.tr/tr/download/article-file/1227592)


### import libraries 

In [3]:
import pandas as pd 
import numpy as np


In [4]:
df = pd.read_excel("Raisin_Dataset.xlsx")
df.head()

,Area,MajorAxisLength,MinorAxisLength,Eccentricity,ConvexArea,Extent,Perimeter,Class
0,87524,442.246011,253.291155,0.819738,90546,0.758651,1184.040,Kecimen
1,75166,406.690687,243.032436,0.801805,78789,0.684130,1121.786,Kecimen
2,90856,442.267048,266.328318,0.798354,93717,0.637613,1208.575,Kecimen
3,45928,286.540559,208.760042,0.684989,47336,0.699599,844.162,Kecimen
4,79408,352.190770,290.827533,0.564011,81463,0.792772,1073.251,Kecimen


In [7]:
df.Class.unique()

array(['Kecimen', 'Besni'], dtype=object)

There are two class ['Kecimen', 'Besni']

In [8]:
df.shape

(900, 8)

There are total 900 records and using all the features that we have available, we will build a classification model by using support vector machine 

### Train_test_split

In [14]:
from sklearn.model_selection import train_test_split 
X = df.drop(["Class"] , axis=1)
y = df["Class"]

x_train , x_test , y_train , y_test   = train_test_split(X , y , test_size=0.3 , random_state=12)
x_train.shape

(630, 7)

### Model Training Using SVM: RBF Kernel

In [25]:
from sklearn.svm import SVC

model = SVC(kernel="rbf")

model.fit(x_train , y_train)

y_predict = model.predict(x_test)

from sklearn.metrics import classification_report
report = classification_report(y_test , y_predict)
print(report)

model.n_iter_

              precision    recall  f1-score   support

       Besni       0.87      0.71      0.78       140
     Kecimen       0.74      0.88      0.81       130

    accuracy                           0.80       270
   macro avg       0.81      0.80      0.80       270
weighted avg       0.81      0.80      0.80       270



array([193], dtype=int32)

### Model Training Using SVM: Linear Kernel

In [26]:
from sklearn.svm import SVC

model = SVC(kernel="linear")

model.fit(x_train , y_train)

y_predict = model.predict(x_test)

from sklearn.metrics import classification_report
report = classification_report(y_test , y_predict)
print(report)

model.n_iter_

              precision    recall  f1-score   support

       Besni       0.83      0.80      0.81       140
     Kecimen       0.79      0.82      0.81       130

    accuracy                           0.81       270
   macro avg       0.81      0.81      0.81       270
weighted avg       0.81      0.81      0.81       270



array([182173827], dtype=int32)

We used two type of kernel but you can simply see the difference of itreations 

### Scale data 

In [33]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(x_train) 

X_train_scaled = scaler.transform(x_train)
X_test_scaled = scaler.transform(x_test)


### Model Training Using SVM: RBF Kernel with scaling 

In [34]:
from sklearn.svm import SVC

model = SVC(kernel="rbf")

model.fit(X_train_scaled , y_train)

y_predict = model.predict(X_test_scaled)

from sklearn.metrics import classification_report
report = classification_report(y_test , y_predict)
print(report)

model.n_iter_

              precision    recall  f1-score   support

       Besni       0.90      0.79      0.84       140
     Kecimen       0.80      0.91      0.85       130

    accuracy                           0.84       270
   macro avg       0.85      0.85      0.84       270
weighted avg       0.85      0.84      0.84       270



array([336], dtype=int32)

As you can see above, after scaling the model performance (accuracy, precision, recall) improves. With scaled data training requires a few more iterations but it is not too big, on the other hand precision, recall, accuracy improvements can be very valuable

### Model Training Using SVM: Linear Kernel with scaling 

In [35]:
from sklearn.svm import SVC

model = SVC(kernel="linear")

model.fit(X_train_scaled , y_train)

y_predict = model.predict(X_test_scaled)

from sklearn.metrics import classification_report
report = classification_report(y_test , y_predict)
print(report)

model.n_iter_

              precision    recall  f1-score   support

       Besni       0.88      0.79      0.83       140
     Kecimen       0.80      0.88      0.84       130

    accuracy                           0.84       270
   macro avg       0.84      0.84      0.84       270
weighted avg       0.84      0.84      0.84       270



array([708], dtype=int32)

For linear kernal, scaling the data reduced the number of iterations drastically (down to 708 from 85005907). The model performance on the other hand is more or less the same. There is some difference but nothing major.

### sk learn pipeline

Its help to reduct the size of code . 

In [42]:
from sklearn.pipeline import Pipeline
pipelines = Pipeline([
        ('scaler', StandardScaler()),
        ('svm', SVC(kernel ="rbf"))
])
pipelines.fit(x_test , y_test )
y_pred = pipelines.predict(x_test)

report = classification_report(y_test , y_pred)
print(report)

              precision    recall  f1-score   support

       Besni       0.90      0.76      0.83       140
     Kecimen       0.78      0.91      0.84       130

    accuracy                           0.83       270
   macro avg       0.84      0.84      0.83       270
weighted avg       0.84      0.83      0.83       270



The sklearn Pipeline combines preprocessing and the model into a single workflow, which keeps the code clean, prevents data leakage, and helps build a production‑ready ML pipeline